### PR TITLE
Accesspoint feature enhancement

### PR DESCRIPTION
**Pull request by md-rafeek**
## Description
34362 - [AP Configuration] : Provision/re-provision AP doesn't work
38737 - Access-Point Configuration: Require module support to perform AP reboot
UT covered 81%

## Type of Change
- [x] Bug fix
- [x] New feature
- [x] Breaking change
- [x] Documentation update

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [x] Tasks are idempotent (can be run multiple times without changing state)
- [ ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [x] Playbooks are modular and reusable
- [x] Handlers are used for actions that need to run on change

## Documentation
- [x] All options and parameters are documented clearly.
- [x] Examples are provided and tested.
- [x] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/accesspoint_workflow_manager.py [hunk 1]
```diff
@@ -516,6 +516,20 @@
           (eg. "mac_address,eth_mac,ap_name,led_brightness_level,led_status,location,radioDTOs")
         type: str
         required: False
+      reboot_ap:
+        description: Reboot one or more access points identified by the MAC Address list.
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:520]: We already discussed in the call. 
Can we change the description?
please add docstring for hostname and ip
Can't we give list of host_names, mac_addresses to reboot ap?
is it the type is list and elements are string?
```
reboot_aps:
  description: Reboot one or more access points (APs) identified by their MAC addresses, hostnames, or management IP addresses.
  type: dict
  required: False
  suboptions:
    mac_addresses:
      description: |
        A list of MAC addresses to identify the access points to be rebooted. 
        At least one of the following parameters is required to identify the APs:
        - mac_addresses
        - hostnames
        - management_ip_addresses
      type: list
      elements: str
      required: True
    hostnames:
      description: |
        A list of hostnames used to identify the access points for rebooting.
      type: list
      elements: str
      required: False
    management_ip_addresses:
      description: |
        A list of management IP addresses used to identify the access points for rebooting.
      type: list
      elements: str
      required: False

```

File: plugins/modules/accesspoint_workflow_manager.py [hunk 2]
```diff
@@ -932,6 +945,44 @@
...
+        dnac_verify: "{{ dnac_verify }}"
+        dnac_port: "{{ dnac_port }}"
+        dnac_version: "{{ dnac_version }}"
+        dnac_debug: "{{ dnac_debug }}"
+        dnac_log: True
+        dnac_log_level: DEBUG
+        config_verify: True
+        state: merged
+        config:
+          - reboot_ap:
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:982]: -  reboot_aps:
        mac_addresses:
          - "AA:BB:CC:DD:EE:FF"
          - "11:22:33:44:55:66"

File: plugins/modules/accesspoint_workflow_manager.py (hunk 1 above)
Comment (md-rafeek) [reply plugins/modules/accesspoint_workflow_manager.py:520]: the same updated. in the yaml document

File: plugins/modules/accesspoint_workflow_manager.py (hunk 2 above)
Comment (md-rafeek) [reply plugins/modules/accesspoint_workflow_manager.py:982]: Updated and changed the code according to this input.

**Review by md-rafeek — COMMENTED** [review]
Both 2 comment addressed

File: plugins/modules/accesspoint_workflow_manager.py [hunk 3]
```diff
@@ -2174,6 +2268,73 @@ def verify_ap_provision(self, wlc_ip_address):
...
+                params=assign_network_device_to_site
+            )
+            self.log("Response device to site: {0} .".format(response["response"]), "INFO")
+
+            if response and isinstance(response, dict):
+                task_id = response.get("response", {}).get("taskId")
+                resync_retry_count = int(self.payload.get("dnac_api_task_timeout"))
+                resync_retry_interval = int(self.payload.get("dnac_task_poll_interval"))
+
+                while resync_retry_count:
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:2305]: resync_retry_count > 0

File: plugins/modules/accesspoint_workflow_manager.py [hunk 4]
```diff
@@ -2174,6 +2268,73 @@ def verify_ap_provision(self, wlc_ip_address):
...
+                    responses = {}
+                    if task_details_response.get("endTime") is not None:
+                        if task_details_response.get("status") == "SUCCESS":
+                            self.log("Task Details: {0} .".format(self.pprint(
+                                task_details_response)), "INFO")
+                            self.msg = "AP {0} Site - {1} updated Successfully".format(
+                                self.have["current_ap_config"].get("ap_name"), self.have.get("site_id"))
+                            self.log(self.msg, "INFO")
+                            return True
+                        else:
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:2317]: else is not required, move the code out of else part.. Anyhow you break if task_details_response.get("endTime") is not None ..

File: plugins/modules/accesspoint_workflow_manager.py [hunk 5]
```diff
@@ -2174,6 +2268,73 @@ def verify_ap_provision(self, wlc_ip_address):
...
+        try:
+            response = self.dnac._exec(
+                family="site_design",
+                function='assign_network_devices_to_a_site',
+                op_modifies=True,
+                params=assign_network_device_to_site
+            )
+            self.log("Response device to site: {0} .".format(response["response"]), "INFO")
+
+            if response and isinstance(response, dict):
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:2300]: check for failure and return..
```

            self.msg = "Failed to receive a valid response from site assignment API."
            self.log(self.msg, "ERROR")
            self.status = "failed"
            self.module.fail_json(msg=self.msg)

```

File: plugins/modules/accesspoint_workflow_manager.py [hunk 6]
```diff
@@ -2839,6 +3138,23 @@ def main():
...
+        ccc_network.validate_ap_config_parameters(ccc_network.validated_config[0]).check_return_status()
+        ap_indentity = list(reboot_list.keys())[0]
+        if ap_indentity and ap_indentity in ccc_network.keymap and len(reboot_list.get(ap_indentity)) > 0:
+            eth_mac_list = []
+            for each_ap in reboot_list[ap_indentity]:
+                ap_indentity_param = {ccc_network.keymap[ap_indentity]: each_ap}
+                ccc_network.log("Reboot AP: {0}".format(str(ap_indentity_param)), "INFO")
+                ap_exist, ap_details = ccc_network.get_accesspoint_details(ap_indentity_param)
+                eth_mac_list.append(ap_details.get("ap_ethernet_mac_address"))
+            ccc_network.reboot_accesspoint(eth_mac_list)
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:3155]:
```
        if eth_mac_list:
            ccc_network.log("Ethernet MAC addresses to reboot: {}".format(eth_mac_list), "INFO")
            ccc_network.reboot_accesspoint(eth_mac_list)
            module.exit_json(**ccc_network.result)
        else:
            ccc_network.log("No valid Ethernet MAC addresses found for reboot.", "WARNING")
else:
    ccc_network.log("No reboot list found in validated configuration.", "WARNING")
```

**Review by md-rafeek — COMMENTED** [review]
addressed all 31 comments

File: plugins/module_utils/dnac.py [hunk 7]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
                 self.log("An error occurred in 'get_sites':{0}".format(e), "ERROR")
                 return None
 
+    def assign_device_to_site(self, device_ids, site_id, site_name=None):
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: removed none.

File: plugins/module_utils/dnac.py [hunk 8]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
...
+
+        Description:
+            Checks if the specified AP assigned to the site returns "True", otherwise logs an error
+            and returns "failed" with error details.
+        """
+        assign_network_device_to_site = {
+            'deviceIds': device_ids,
+            'siteId': site_id,
+        }
+        self.log("Assign device to site before update: {0}, {1} ."
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: updated

File: plugins/module_utils/dnac.py [hunk 9]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
...
+                 .format(site_name, assign_network_device_to_site), "INFO")
+        try:
+            response = self.dnac._exec(
+                family="site_design",
+                function='assign_network_devices_to_a_site',
+                op_modifies=True,
+                params=assign_network_device_to_site
+            )
+            self.log("Response from assigning device to site: {0}, {1}, {2} .".format(
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: updated

File: plugins/module_utils/dnac.py [hunk 10]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
...
+                        self.msg = "Unable to get success response, hence AP site not updated"
+                        self.log(self.msg, "ERROR")
+                        self.log("Task Details: {0} .".format(self.pprint(
+                            task_details_response)), "ERROR")
+                        responses["accesspoints_updates"] = {
+                            "ap_assign_to_site_task_details": task_details_response,
+                            "ap_assign_to_site_status": self.msg}
+                        self.module.fail_json(msg=self.msg, response=responses)
+                        break
+                    time.sleep(resync_retry_interval)
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py:817]: updated

File: plugins/module_utils/dnac.py [hunk 11]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
...
+                            self.log("Task Details: {0} .".format(self.pprint(
+                                task_details_response)), "INFO")
+                            self.msg = "AP {0} Site - {1} updated Successfully".format(
+                                self.have["current_ap_config"].get("ap_name"), self.have.get("site_id"))
+                            self.log(self.msg, "INFO")
+                            return True
+
+                        self.result['changed'] = True if self.result['changed'] is True else False
+                        self.status = "failed"
+                        self.msg = "Unable to get success response, hence AP site not updated"
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: removed AP

File: plugins/module_utils/dnac.py [hunk 12]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
...
+                site_name, str(assign_network_device_to_site), str(response["response"])), "INFO")
+
+            if response and isinstance(response, dict):
+                task_id = response.get("response", {}).get("taskId")
+                resync_retry_count = int(self.payload.get("dnac_api_task_timeout"), 100)
+                resync_retry_interval = int(self.payload.get("dnac_task_poll_interval"), 5)
+
+                while resync_retry_count > 0:
+                    task_details_response = self.get_tasks_by_id(task_id)
+                    self.log("Status of the task: {0} .".format(self.status), "INFO")
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: removed the log

File: plugins/module_utils/dnac.py [hunk 13]
```diff
@@ -749,6 +749,80 @@ def get_site(self, site_name):
...
+                family="site_design",
+                function='assign_network_devices_to_a_site',
+                op_modifies=True,
+                params=assign_network_device_to_site
+            )
+            self.log("Response from assigning device to site: {0}, {1}, {2} .".format(
+                site_name, str(assign_network_device_to_site), str(response["response"])), "INFO")
+
+            if response and isinstance(response, dict):
+                task_id = response.get("response", {}).get("taskId")
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py:782]: updated

**Review by md-rafeek — COMMENTED** [review]
Addressed all 7 comments

File: plugins/module_utils/dnac.py [hunk 14]
```diff
@@ -749,6 +749,86 @@ def get_site(self, site_name):
                 self.log("An error occurred in 'get_sites':{0}".format(e), "ERROR")
                 return None
 
+    def assign_device_to_site(self, device_ids, site_id, site_name):
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: Modified

File: plugins/module_utils/dnac.py [hunk 15]
```diff
@@ -749,6 +749,86 @@ def get_site(self, site_name):
...
+        Assign the Device to the Site
+
+        Parameters:
+            self (object): An instance of a class used for interacting with Cisco Catalyst Center.
+
+        Returns:
+            bool: True if The device assign to the site, otherwise False.
+
+        Description:
+            Checks if the specified AP assigned to the site returns "True", otherwise logs an error
```
[the remark this answers is not in the document]
Comment (md-rafeek) [reply plugins/module_utils/dnac.py]: updated changed to devices

**Review by md-rafeek — COMMENTED** [review]
Addressed 2 comments

**Review by madhansansel — APPROVED** [review]

**Pull request merged**